win32ss/user/user32/misc/exticon.c

```diff
@@ -714,9 +714,75 @@ static UINT ICO_ExtractIconExW(
 	    xresdir = find_entry_by_id(iconresdir, LOWORD(pIconId[i]), rootresdir);
             if( !xresdir )
             {
+#ifdef __REACTOS__
+              const IMAGE_RESOURCE_DIRECTORY *resdir;
+              WARN("icon entry %d not found\n", LOWORD(pIconId[i]));
+
+              /* Try to get an icon by walking the files Resource Section */
+              if (iconresdir->NumberOfIdEntries > 0)
+              {
+                  xresent = (const IMAGE_RESOURCE_DIRECTORY_ENTRY*)
+                      (ULONG_PTR)(iconresdir + 1);
+              }
+              else
+              {
+                  RetPtr[i] = 0;
+                  continue;
+              }
+
+              /* Get the Resource Directory */
+              resdir = (const IMAGE_RESOURCE_DIRECTORY *)
+                       ((const char *)rootresdir + xresent->OffsetToDirectory);
+
+              if (resdir->NumberOfIdEntries > 0) // Do we have entries
+              {
+                  xresent = (const IMAGE_RESOURCE_DIRECTORY_ENTRY*)
+                      (ULONG_PTR)(resdir + 1);
+              }
+              else
+              {
+                  RetPtr[i] = 0;
+                  continue;
+              }
+
+              /* Retrieve the data entry and find its address */
+              igdataent = (const IMAGE_RESOURCE_DATA_ENTRY*)
+                  ((const char *)rootresdir + xresent->OffsetToData);
+              idata = RtlImageRvaToVa(RtlImageNtHeader((HMODULE)peimage),
+                  (HMODULE)peimage, igdataent->OffsetToData, NULL);
+              if (!idata)
+              {
+                  RetPtr[i] = 0;
+                  continue;
+              }
+
+              /* Check to see if this looks like an icon bitmap */
+              if (idata[0] == sizeof(BITMAPINFOHEADER))
+              {
+                  BITMAPINFOHEADER bmih;
+                  RtlCopyMemory(&bmih, idata, sizeof(BITMAPINFOHEADER));
+                  /* Do the Width and Height look correct for an icon */
+                  if ((bmih.biWidth * 2) == bmih.biHeight)
+                  {
+                      RetPtr[0] = CreateIconFromResourceEx(idata, igdataent->Size,
+                          TRUE, 0x00030000, cx1, cy1, flags);
+                      if (cx2 && cy2)
+                          RetPtr[1] = CreateIconFromResourceEx(idata, idataent->Size,
+                              TRUE, 0x00030000, cx2, cy2, flags);
+                      ret = 1;  // Set number of icons found
+                      goto end; // Success so Exit
+                  }
+              }
+              else
+              {
+                  RetPtr[i] = 0;
+                  continue;
+              }
+#else
               WARN("icon entry %d not found\n", LOWORD(pIconId[i]));
 	      RetPtr[i]=0;
 	      continue;
+#endif
```

Comment:
What I can see here is looking more like a workaround. It's a big code block added at the failure case when the icon is not found. I think there should be a smaller proper fix targetting the code above, or the functions it uses, effectively making it not fail in this case.

Reply:
Yes, I also think it is a workaround for a (rare) corner case.
Suggestion:
1.) Add a 2k3sp2 pic at least that proves our target displaying that specific icon.
2.) Since it is likely that we will change this code again in the future due to it being a workaround, it would be good to have an automated test on the testbots which covers this case.
If both are met, I will give my Approval for committing that workaround.

Reply:
@binarymaster and @JoachimHenze, each of you are certainly entitled to your opinion on whether this is a workaround. But, after working with this for quite some hours, I disagree. If you look at this file in CFF Explorer, which I have provided below, you will see that to check for the correctness of the 'Icon Group' tree, you must navigate the lower boxed area shown. Once you find that this contains inconsistent data, to get to the 'Icons' you must navigate the upper boxed area as shown below. Thanks very much to each of you for reviewing and providing input here.

![F197_Resource_Directory](https://github.com/reactos/reactos/assets/32620577/5f3d7496-628c-4a93-8ea6-918dce8db661)

Reply:
We do better understand now.
The specific F197I.exe built with the WatCom C++ compiler is most likely not entirely legit.
Due to your further analysis we do know now that:
WinXP and 2k3sp2 do manage to display that icon in their explorer,
while Win7 shows only a blank icon.
I am aware that you wrote a testcase for that scenario, where you put all the information into also within a comment.
I will approve committing your test-case, and I will then afterwards also approve your fix.
Due to it being such a tricky corner case, I want to thank you again for having invested so much love in the analysis, in the test and in the fix.

Reply:
Thanks for the positive comments. You are very welcome. I enjoy the challenges.